openstack/data_source_openstack_identity_projects_v3.go

```diff
@@ -0,0 +1,160 @@
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+
+			"enabled": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+			},
+
+			"is_domain": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+
+			"name": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"name_regex"},
+			},
+
+			"name_regex": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				ValidateFunc:  validation.StringIsValidRegExp,
+				ConflictsWith: []string{"name"},
+			},
+
+			"parent_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"tags": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Set:      schema.HashString,
+			},
+
+			// Computed values
+			"ids": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+		},
+	}
+}
+
+// dataSourceIdentityProjectIdsV3Read performs the project lookup.
+func dataSourceIdentityProjectIdsV3Read(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*Config)
+
+	identityClient, err := config.IdentityV3Client(GetRegion(d, config))
+	if err != nil {
+		return diag.Errorf("Error creating OpenStack identity client: %s", err)
+	}
+
+	enabled := d.Get("enabled").(bool)
+	isDomain := d.Get("is_domain").(bool)
+
+	tags := []string{}
+	tagList := d.Get("tags").(*schema.Set).List()
+	for _, v := range tagList {
+		tags = append(tags, fmt.Sprint(v))
+	}
+	joinedTags := strings.Join(tags, ",")
+
+	listOpts := projects.ListOpts{
+		DomainID: d.Get("domain_id").(string),
+		Enabled:  &enabled,
+		IsDomain: &isDomain,
+		Name:     d.Get("name").(string),
+		ParentID: d.Get("parent_id").(string),
+		Tags:     joinedTags,
+	}
+
+	allPages, err := projects.List(identityClient, listOpts).AllPages()
+	if err != nil {
+		return diag.Errorf("Unable to list projects in openstack_identity_project_ids_v3: %s", err)
+	}
+
+	allProjects, err := projects.ExtractProjects(allPages)
+	if err != nil {
+		return diag.Errorf("Unable to retrieve projects in openstack_identity_project_ids_v3: %s", err)
+	}
+
+	log.Printf("[DEBUG] Retrieved %d images in openstack_identity_project_ids_v3: %+v", len(allProjects), allProjects)
+
+	nameRegex, nameRegexOk := d.GetOk("name_regex")
+	if nameRegexOk {
+		allProjects = projectsFilterByRegex(allProjects, nameRegex.(string))
+		log.Printf("[DEBUG] Project list filtered by regex: %s", d.Get("name_regex"))
+	}
+
+	log.Printf("[DEBUG] Got %d projects after filtering in openstack_identity_project_ids_v3: %+v", len(allProjects), allProjects)
+
+	projectIDs := make([]string, len(allProjects))
+	for i, image := range allProjects {
+		projectIDs[i] = image.ID
+	}
+
+	d.SetId(fmt.Sprintf("%d", hashcode.String(strings.Join(projectIDs, ","))))
```

Comment:
I think the below code should be enough, since it's already a string:
```suggestion
	d.SetId(hashcode.String(strings.Join(projectIDs, ",")))
```

Reply:
hashcode.String() returns an int:
```func hashcode.String(s string) int
String hashes a string to a unique hashcode.

crc32 returns a uint32, but for our use we need and non negative integer. Here we cast to an integer and invert it if the result is negative.```

Reply:
Interesting, you don't expect to get `int` from the `String()` method.